File: order-summary-tax-excluded-bogus-refund/python/detect_tax_dropped_from_summary.py
"""Flag Medusa v2 orders whose summary reports a phantom overpayment because
tax_total was left out of the summary's own totals math. order.total is
computed correctly as subtotal + shipping_total + tax_total minus discounts,
but summary.accounting_total and the pending_difference built on top of it
are computed from subtotal + shipping_total alone (see medusajs/medusa#13405).
A fully paid, tax-inclusive order therefore looks overpaid by exactly the
tax amount, and anything wired to pending_difference can issue a bogus
refund for money nobody actually overpaid. This script only flags the
divergence, and if a refund already matched the missing tax it flags that
too, gated behind DRY_RUN. It never issues a refund or a recharge. Safe to
run again and again.

Guide: https://www.allanninal.dev/medusa/order-summary-tax-excluded-bogus-refund/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_tax_dropped_from_summary")

BACKEND_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
ADMIN_EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
ADMIN_PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

EPSILON = 0.01

ORDER_FIELDS = (
    "id,display_id,total,tax_total,subtotal,shipping_total,"
    "summary.accounting_total,summary.current_order_total,summary.pending_difference,"
    "summary.paid_total,summary.transaction_total,summary.refunded_total"
)


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def admin_get(token, path, params=None):
    r = requests.get(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def admin_post(token, path, json_body):
    r = requests.post(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        json=json_body,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def detect_tax_dropped_from_summary(order, epsilon=EPSILON):
    """Pure decision function. No I/O.

    order: {
      "total": float, "tax_total": float,
      "summary": {"accounting_total": float, "current_order_total": float,
                  "pending_difference": float, "paid_total": float},
    }

    Returns {"affected", "drift", "correctedPendingDifference"}.
    """
    summary = order["summary"]
    drift = order["total"] - summary["accounting_total"]
    tax_total = order["tax_total"]

    affected = tax_total > 0 and abs(drift - tax_total) <= epsilon
    corrected_pending_difference = order["total"] - summary["paid_total"]

    return {
        "affected": affected,
        "drift": drift,
        "correctedPendingDifference": corrected_pending_difference,
    }


def already_refunded_tax(order_id, tax_total, token, epsilon=EPSILON):
    data = admin_get(
        token,
        f"/admin/orders/{order_id}/payment-collections",
        {"fields": "id,status,*payments,*payments.refunds"},
    )
    for collection in data.get("payment_collections", []):
        for payment in collection.get("payments") or []:
            for refund in payment.get("refunds") or []:
                if abs(refund.get("amount", 0) - tax_total) <= epsilon:
                    return True
    return False


def flag_order(order_id, tax_total, token):
    admin_post(token, f"/admin/orders/{order_id}", {
        "metadata": {"flagged_tax_refund_drift": True, "expected_manual_recharge": tax_total},
    })


def list_orders_with_summary(token):
    orders = []
    offset = 0
    limit = 100
    while True:
        data = admin_get(token, "/admin/orders", {
            "fields": ORDER_FIELDS,
            "limit": limit,
            "offset": offset,
        })
        orders.extend(data["orders"])
        offset += limit
        if offset >= data["count"]:
            return orders


def run():
    token = get_admin_token()
    raw_orders = list_orders_with_summary(token)

    report = {}
    flagged = 0
    for raw_order in raw_orders:
        outcome = detect_tax_dropped_from_summary(raw_order)
        if not outcome["affected"]:
            continue

        order_id = raw_order["id"]
        report[order_id] = {
            "correct_total": raw_order["total"],
            "buggy_accounting_total": raw_order["summary"]["accounting_total"],
            "drift": raw_order["tax_total"],
        }

        already_refunded = already_refunded_tax(order_id, raw_order["tax_total"], token)
        flagged += 1
        log.warning(
            "Order %s tax dropped from summary: correct_total=%s buggy_accounting_total=%s "
            "drift=%s corrected_pending_difference=%s already_refunded=%s. %s",
            raw_order.get("display_id") or order_id,
            raw_order["total"], raw_order["summary"]["accounting_total"],
            outcome["drift"], outcome["correctedPendingDifference"], already_refunded,
            "would flag for review" if DRY_RUN else "flagging for review",
        )
        if already_refunded and not DRY_RUN:
            flag_order(order_id, raw_order["tax_total"], token)

    log.info(
        "Done. %d order(s) with tax dropped from the summary flagged for manual review. "
        "No refund or recharge was issued by this script.",
        flagged,
    )
    return report


if __name__ == "__main__":
    run()

File: order-summary-tax-excluded-bogus-refund/node/detect-tax-dropped-from-summary.js
/**
 * Flag Medusa v2 orders whose summary reports a phantom overpayment because
 * tax_total was left out of the summary's own totals math. order.total is
 * computed correctly as subtotal + shipping_total + tax_total minus discounts,
 * but summary.accounting_total and the pending_difference built on top of it
 * are computed from subtotal + shipping_total alone (see medusajs/medusa#13405).
 * A fully paid, tax-inclusive order therefore looks overpaid by exactly the
 * tax amount, and anything wired to pending_difference can issue a bogus
 * refund for money nobody actually overpaid. This script only flags the
 * divergence, and if a refund already matched the missing tax it flags that
 * too, gated behind DRY_RUN. It never issues a refund or a recharge. Safe to
 * run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/order-summary-tax-excluded-bogus-refund/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const EPSILON = 0.01;

const ORDER_FIELDS =
  "id,display_id,total,tax_total,subtotal,shipping_total," +
  "summary.accounting_total,summary.current_order_total,summary.pending_difference," +
  "summary.paid_total,summary.transaction_total,summary.refunded_total";

export function detectTaxDroppedFromSummary(order, epsilon = EPSILON) {
  // Pure: no I/O. order = { total, tax_total,
  // summary: { accounting_total, current_order_total, pending_difference, paid_total } }
  const { summary } = order;
  const drift = order.total - summary.accounting_total;
  const taxTotal = order.tax_total;

  const affected = taxTotal > 0 && Math.abs(drift - taxTotal) <= epsilon;
  const correctedPendingDifference = order.total - summary.paid_total;

  return { affected, drift, correctedPendingDifference };
}

async function getAdminToken() {
  const res = await fetch(`${BACKEND_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: ADMIN_EMAIL, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function adminGet(token, path, params = {}) {
  const url = new URL(`${BACKEND_URL}${path}`);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${path} ${res.status}`);
  return res.json();
}

async function adminPost(token, path, jsonBody) {
  const res = await fetch(`${BACKEND_URL}${path}`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify(jsonBody),
  });
  if (!res.ok) throw new Error(`Medusa ${path} ${res.status}`);
  return res.json();
}

async function alreadyRefundedTax(orderId, taxTotal, token, epsilon = EPSILON) {
  const data = await adminGet(token, `/admin/orders/${orderId}/payment-collections`, {
    fields: "id,status,*payments,*payments.refunds",
  });
  for (const collection of data.payment_collections || []) {
    for (const payment of collection.payments || []) {
      for (const refund of payment.refunds || []) {
        if (Math.abs((refund.amount || 0) - taxTotal) <= epsilon) return true;
      }
    }
  }
  return false;
}

async function flagOrder(orderId, taxTotal, token) {
  await adminPost(token, `/admin/orders/${orderId}`, {
    metadata: { flagged_tax_refund_drift: true, expected_manual_recharge: taxTotal },
  });
}

async function listOrdersWithSummary(token) {
  const orders = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const data = await adminGet(token, "/admin/orders", {
      fields: ORDER_FIELDS,
      limit,
      offset,
    });
    orders.push(...data.orders);
    offset += limit;
    if (offset >= data.count) return orders;
  }
}

export async function run() {
  const token = await getAdminToken();
  const rawOrders = await listOrdersWithSummary(token);

  const report = {};
  let flagged = 0;
  for (const rawOrder of rawOrders) {
    const outcome = detectTaxDroppedFromSummary(rawOrder);
    if (!outcome.affected) continue;

    const orderId = rawOrder.id;
    report[orderId] = {
      correct_total: rawOrder.total,
      buggy_accounting_total: rawOrder.summary.accounting_total,
      drift: rawOrder.tax_total,
    };

    const alreadyRefunded = await alreadyRefundedTax(orderId, rawOrder.tax_total, token);
    flagged++;
    console.warn(
      `Order ${rawOrder.display_id || orderId} tax dropped from summary: ` +
        `correct_total=${rawOrder.total} buggy_accounting_total=${rawOrder.summary.accounting_total} ` +
        `drift=${outcome.drift} corrected_pending_difference=${outcome.correctedPendingDifference} ` +
        `already_refunded=${alreadyRefunded}. ` +
        `${DRY_RUN ? "would flag for review" : "flagging for review"}`
    );
    if (alreadyRefunded && !DRY_RUN) await flagOrder(orderId, rawOrder.tax_total, token);
  }

  console.log(
    `Done. ${flagged} order(s) with tax dropped from the summary flagged for manual review. ` +
      `No refund or recharge was issued by this script.`
  );
  return report;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
